---
title: "Clean dataset"
format: html
---

# Load libraries

```{r}
#| message: false
#| warning: false
library("tidyverse")
```

# Data preparation

Using the `str` command to investigate, what the data looks like:

```{r}
str(genome_metadata)
```

The data consist of a dataframe with 618 observations and 20 variables. We recognize that observations with no value found has the identifier "not provided". Additionally one variable contains phylogenetic information that can be split into different variable containing information about the different classes of the evolutionary history.

## Split data

We want to split the data in the "lineage" variable into multiple variables to store the information about each of the phylogenetic classes individually. First we investigate the data in the lineage variable:

```{r}
genome_metadata |>
  select(Lineage) |>
  slice_sample(n = 10)
```

By investigating the data it is seen that each string of phylogenetic information consist of identifiers for each of the different classes in the taxonomic rank. These identifiers are:

-   d = domain

-   p = phylum

-   c = class

-   o = order

-   f = family

-   g = genus

-   s = species

We then split the data into the individual taxonomy class ranks.

```{r}
genome_metadata_clean <- genome_metadata |>
  mutate(
    Domain = case_when(str_detect(Lineage, "d__") ~ str_remove(str_extract(Lineage, "d__[^;]*"), "d__")),
    Phylum = case_when(str_detect(Lineage, "p__") ~ str_remove(str_extract(Lineage, "p__[^;]*"), "p__")),
    Class  = case_when(str_detect(Lineage, "c__") ~ str_remove(str_extract(Lineage, "c__[^;]*"), "c__")),
    Order  = case_when(str_detect(Lineage, "o__") ~ str_remove(str_extract(Lineage, "o__[^;]*"), "o__")),
    Family = case_when(str_detect(Lineage, "f__") ~ str_remove(str_extract(Lineage, "f__[^;]*"), "f__")),
    Genus  = case_when(str_detect(Lineage, "g__") ~ str_remove(str_extract(Lineage, "g__[^;]*"), "g__")),
    Species = case_when(str_detect(Lineage, "s__") ~ str_remove(str_extract(Lineage, "s__[^;]*"), "s__"))
  )
```

*Another suggestion for how the data can be split. Also added a code chunk that writes NA if there is no data*

```{r}
genome_metadata_clean_2 <- genome_metadata |>
  mutate(
    # Split the Lineage column into 7 taxonomic ranks
    split_lineage = str_split_fixed(Lineage, ";", 7),

    Domain  = str_remove(split_lineage[, 1], "d__"),
    Phylum  = str_remove(split_lineage[, 2], "p__"),
    Class   = str_remove(split_lineage[, 3], "c__"),
    Order   = str_remove(split_lineage[, 4], "o__"),
    Family  = str_remove(split_lineage[, 5], "f__"),
    Genus   = str_remove(split_lineage[, 6], "g__"),
    Species = str_remove(split_lineage[, 7], "s__")
  ) |>
  mutate(
    across(
      c(Domain, Phylum, Class, Order, Family, Genus, Species),
      ~ na_if(.x, "")
    )
  )|>
  select(-split_lineage) # removing the helper column
```

## Missing values
