---
title: "Clean dataset"
format: html
---

# Clean Data

## 1. Load libraries

```{r}
#| message: false
#| warning: false
library("tidyverse")
```

## 2. Load data

Loading the data saved in the 01_dat_load.tsv file

```{r}
genome_metadata <- readr::read_tsv("../data/01_dat_load.tsv")
```

## 3. Data preparation

Using the `str` command to investigate, what the data looks like:

```{r}
str(genome_metadata)
```

The data consist of a data-frame with 618 observations and 20 variables. We recognize that observations with no value found has the identifier "not provided". Additionally one variable contains phylogenetic information that can be split into different variable containing information about the different classes of the evolutionary history.

### 3.1 Split data

We want to split the data in the "lineage" variable into multiple variables to store the information about each of the phylogenetic classes individually. First we investigate the data in the lineage variable:

```{r}
genome_metadata |>
  dplyr::select(Lineage) |>
  dplyr::slice_sample(n = 10)
```

By investigating the data it is seen that each string of phylogenetic information consist of identifiers for each of the different classes in the taxonomic rank. These identifiers are:

-   d = domain

-   p = phylum

-   c = class

-   o = order

-   f = family

-   g = genus

-   s = species

We want to start with converting all "not provided" to NA across all character columns. Then we want to split the "lineage" variable into its seven taxonomic ranks to store the information about each of the phylogenetic classes individually. Then we want to extract each taxonomic rank and remove prefixes. And convert empty strings to NA in the new taxonomy columns.

We then remove the GTDB suffixes (e.g. "\_A") to streamline taxonomies. Lastly we remove columns that will not be used in our analysis.

```{r}
genome_metadata_clean <- genome_metadata |>
  dplyr::mutate(
    dplyr::across(
      .cols = where(is.character),
      .fns  = ~ dplyr::na_if(.x, "not provided")
    )
  ) |>
  dplyr::mutate(
    split_lineage = stringr::str_split_fixed(Lineage, ";", 7),

    Domain  = stringr::str_remove(split_lineage[, 1], "d__"),
    Phylum  = stringr::str_remove(split_lineage[, 2], "p__"),
    Class   = stringr::str_remove(split_lineage[, 3], "c__"),
    Order   = stringr::str_remove(split_lineage[, 4], "o__"),
    Family  = stringr::str_remove(split_lineage[, 5], "f__"),
    Genus   = stringr::str_remove(split_lineage[, 6], "g__"),
    Species = stringr::str_remove(split_lineage[, 7], "s__")
  ) |>
  dplyr::mutate(
    dplyr::across(
      .cols = c(Domain, Phylum, Class, Order, Family, Genus, Species),
      .fns  = ~ dplyr::na_if(.x, "")
    )
  ) |>
  dplyr::mutate(
    dplyr::across(
      .cols = Phylum:Species,
      .fns  = ~ stringr::str_remove_all(as.character(.x), "_[A-Za-z0-9]+")
    )
  ) |>
  dplyr::select(
    -split_lineage,
    -Genome_accession,
    -Species_rep,
    -Lineage,
    -Sample_accession,
    -Study_accession,
    -FTP_download
  )
```

### 3.2 Viewing the data

Viewing the top 10 rows of the split data to see the new structure.

```{r}
genome_metadata_clean |> 
  dplyr::select(Genome, Domain:Species) |> 
  dplyr::slice_head(n = 10)
```

```{r}
view(genome_metadata_clean)
```

### 3.3 Saving the data

Saving the cleaned data into 02_dat_clean.tsv file located in the data directory.

```{r}
readr::write_tsv(
  x    = genome_metadata_clean,
  file = "../data/02_dat_clean.tsv"
)
```
