---
title: "08_analysis_4"
format: html
---

# Loading libraries

```{r}
#| message: false
#| warning: false

library(tidyverse)
library(pheatmap)
library(broom)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(stats)

source("99_proj_func.R")
```

# Preparing dataset

## Removing NA values for "Countries"

From the augmented data, we'll remove the genomes without country information.

```{r}
#| echo: true
#| eval: false
genomes_country <- genome_metadata_aug |> 
  dplyr::filter(!is.na(Country))
```

# About the functions

Functions used in this script were sourced from the file "99_proj_func.R".

## normalize()

The function "normalize()" is used for producing a heatmap and doing PCAs. It normalizes the counts of a variable tree_group based on the sample size of the countries to allow for easy production of several heatmaps. The function works by:

-   Counting the instances of each group per country.

-   Normalizing this count based on the sample size of that country.

-   And converting the data into wide format for easier analysis.

## heatmap_abundance_per_country()

A function to produce the heatmap based on a group in the lineage. First, the data is normalized by sample size for each country; then it is converted into a matrix, and columns with zero variance are removed. The function then produces a heatmap using the pheatmap function. Again, the function was established to be able to quickly produce heatmaps for several tree_groups.

# Heatmaps for exploration of abundance among countries

## Creating heatmaps

Using the defined heatmap_abundancy_per_country() function to produce heatmaps for some of the "higher" lineage groups, Family, Order, and Phylum, to ensure clearer plots.

```{r}
#| echo: true
#| eval: true

heatmap_family <- heatmap_abundance_per_country(tree_group = "Family")
heatmap_order <- heatmap_abundance_per_country(tree_group = "Order")
heatmap_phylum <- heatmap_abundance_per_country(tree_group = "Phylum")
```

## Saving results

The resulting heatmaps are saved as png's to the results directory with appropriate sizes.

```{r}
#| echo: false
#| eval: false

ggplot2::ggsave(
  filename = "../results/Heatmap_family.png",
  plot = heatmap_family,
  width = 190,
  height = 100,
  units = "mm",
  dpi = 300
)
ggplot2::ggsave(
  filename = "../results/Heatmap_order.png",
  plot = heatmap_order,
  width = 190,
  height = 100,
  units = "mm",
  dpi = 300
)
ggplot2::ggsave(
  filename = "../results/Heatmap_phylum.png",
  plot = heatmap_phylum,
  width = 190,
  height = 100,
  units = "mm",
  dpi = 300
)
```

# Principal Component Analysis

To investigate if there is a difference in the distribution of microbiomes between countries, a Prinicipal Component Analysis was carried out using the PCA_abundancy_per_country() function. To do this, the data was normalized based on the sample size for each country, after which it was analysed with the prcomp() function. From this fit, the results were extracted (eigenvalues and rotation matrix) and used to create a biplot of the loadings (p1) and determine the variance explained (plotted in p2). Both plots were saved to the results directory.

Running the function for Phylum, which has fewer variables, hence resulting in a clearer plot.

## Running PCA

To run the PCA, the data was normalized by sample size of each country with the normalize() function based on the selected tree_group. In this case, "Phylum" was chosen to ensure a clearer plot. The data was then analysed with the prcomp() function and from the resulting fit, the eigenvalues and rotation matrix were extracted.

```{r}
#| echo: false
#| eval: true

tree_group_PCA = "Phylum"

group_freq <- normalize(data = genomes_country,
                        tree_group = tree_group_PCA)

country_names <- group_freq$Country

pca_fit <- group_freq |> 
  dplyr::select(-Country) |> 
  scale() |> 
  stats::prcomp()

variance_explained <- pca_fit |> 
  broom::tidy(matrix = "eigenvalues")

rotation_matrix <- pca_fit |> 
  broom::tidy(matrix = "rotation")
```

## Create Biplot

The biplot was created plotting each countries "scores" for the first two principal components. In the plot, the loadings, extracted from the rotation_matrix, were included to show the variance in species abundance between these countries.

```{r}
pca_scores <- pca_fit |> 
  broom::augment(
    group_freq |>
      dplyr::select(-Country)
    ) |> 
  mutate(Country = country_names)

pc1_var <- round(variance_explained$percent[1] * 100, 1)
pc2_var <- round(variance_explained$percent[2] * 100, 1)

loadings_biplot <- rotation_matrix |> 
  dplyr::filter(PC %in% c(1, 2)) |> 
  tidyr::pivot_wider(
    names_from = PC,
    values_from = value,
    names_prefix = "PC"
    ) |> 
  dplyr::mutate(
    PC1 = PC1 * 3,
    PC2 = PC2 * 3)


p1 <- ggplot() +
  ggplot2::geom_point(
    data = pca_scores,
    mapping = ggplot2::aes(x = .fittedPC1,
                           y = .fittedPC2,
                           color = Country),
    size = 4,
    show.legend = TRUE
    ) +
  ggplot2::geom_text(
    data = pca_scores,
    mapping = ggplot2::aes(x = .fittedPC1,
                 y = .fittedPC2,
                 label = Country,
                 color = Country),
    vjust = -1,
    size = 4,
    show.legend = FALSE
    ) +
  ggplot2::geom_segment(
    data = loadings_biplot,
    mapping = ggplot2::aes(x = 0,
                 y = 0,
                 xend = PC1,
                 yend = PC2),
    arrow = arrow(length = unit(0.3, "cm")),
    color = "gray50",
    linewidth = 0.7
    ) +
  ggrepel::geom_text_repel(
    data = loadings_biplot,
    mapping = ggplot2::aes(
      x = PC1,
      y = PC2,
      label = column),
    color = "gray10",
    ize = 3.5,
    fontface = "italic",
    box.padding = 0.5,
    point.padding = 0.3,
    segment.color = "gray60",
    segment.size = 0.3,
    max.overlaps = Inf
    ) +
  ggplot2::labs(
    title = paste0("PCA Biplot: Countries + ", tree_group_PCA),
    x = paste0("PC1 (", pc1_var, "%)"),
    y = paste0("PC2 (", pc2_var, "%)")
    ) +
  ggplot2::theme(
    legend.position = "right",
    plot.title = element_text(face = "bold")
    )

p1
```

## Create plot of variance explained by the PCs

To visualize how much variance was explained by the principal components, a barplot was created using the eigenvalues of the first three principal components.

```{r}
p2 <- ggplot2::ggplot(
  data = variance_explained, 
  mapping = ggplot2::aes(x = PC, y = percent)
  ) +
  ggplot2::geom_col(fill = "steelblue",
                    alpha = 0.8) +
  ggplot2::scale_y_continuous(labels = scales::percent_format()) +
  ggplot2::labs(
    title = "Variance Explained by Each PC",
    x = "Principal Component",
    y = "Variance Explained (%)"
    ) +
  ggplot2::theme(plot.title = element_text(face = "bold"))

p2
```

## Save plots

Using ggsave to save plots in the results directory.

```{r}
ggplot2::ggsave(
  filename = "../results/PCA_biplot.png",
  plot = p1,
  width = 250,
  height = 160,
  units = "mm")
  
ggplot2::ggsave(
  filename = "../results/PCA_variance_explained.png",
  plot = p2,
  width = 110,
  height = 90,
  units = "mm")
```

# Tests: Species' distribution between counties

## Chi-squared test, with permutation.

Test if the distribution of species differs between countries. The groups abundancy is counted using the counting_species() function. We use simulate.p.value = TRUE to compute the p-value via Monte Carlo simulation, which is more reliable when some expected counts are small or zero.

### For Species

```{r}
#| echo: false
#| eval: true

data <- counting_species(data = genomes_country,
                         tree_group = "Species")

tbl <- as.matrix(data[ , -1])
rownames(tbl) <- data$Country

stats::chisq.test(
  x = tbl,
  simulate.p.value = TRUE,
  B = 1e6)

```

The difference in distribution of species between countries is not significant at all, as the p-value is 1.

### For Family

```{r}
#| echo: false
#| eval: true

data <- counting_species(data = genomes_country,
                         tree_group = "Family")

tbl <- as.matrix(data[ , -1])
rownames(tbl) <- data$Country

stats::chisq.test(
  x = tbl,
  simulate.p.value = TRUE,
  B = 1e6)
```

The difference in distribution of species between countries is not significant at all, as the p-value is 0.8.

### For Phylum

```{r}
#| echo: false
#| eval: true

data <- counting_species(data = genomes_country,
                         tree_group = "Phylum")

tbl <- as.matrix(data[ , -1])
rownames(tbl) <- data$Country

stats::chisq.test(
  x = tbl,
  simulate.p.value = TRUE,
  B = 1e6)

```

The difference in distribution of species between countries is not significant, as the p-value is 0.6

# Endometriosis test

To assess whether genes associated with endometriosis are significantly more represented in one country than another, a Fisher’s exact test is performed to compare the frequency of these genes. The Fisher test is chosen because the number of endometriosis-associated genomes is low and the sample sizes between countries are highly uneven, especially for Denmark. Fisher’s exact test can handle both sparse data and unequal group sizes.

The purpose of the test is to answer:

**Does the proportion of endometriosis-associated genomes differ between countries?**

Countries are compared pairwise. For each comparison, the null hypothesis states that the proportion of endometriosis-associated genomes is the same in both countries, while the alternative hypothesis states that the proportions differ.

Before running the test, we confirm that the required variables are present:

```{r}
#| echo: false
#| eval: false

names(genomes_country)
```

Both the `Country` and `endometriosis_associated` variables are present and we are able to make the analysis.

As we have to run the fisher exact test between the three country combinations, the test is made into a function. The columns that the functions use the made variable, so the function can be used to make a fisher exact test comparing the country with other variables, if we wanted to do that.:

```{r}
compute_fisher_results <- function(data, country_var, outcome_var, pairs) {
  
  # Run Fisher test for each pair
  results <- lapply(pairs, function(p) {
    fisher_compare_countries(
      data        = data,
      country_var = country_var,
      outcome_var = outcome_var,
      country1    = p[1],
      country2    = p[2]
    )
  })
  
  # Convert each result into a tibble row
  rows <- lapply(results, function(res) {
    tibble::tibble(
      country1   = res$countries[1],
      country2   = res$countries[2],
      p_value    = round(res$p_value, 4),
      odds_ratio = round(as.numeric(res$odds_ratio), 3),
      CI_low     = round(res$conf_int[1], 3),
      CI_high    = round(res$conf_int[2], 3)
    )
  })
  
  # Combine into a table
  dplyr::bind_rows(rows)
}

```

We then define the country pairs that we want to compare:

```{r}
pairs <- list(
  c("USA", "China"),
  c("USA", "Denmark"),
  c("Denmark", "China")
)
```

The functions is then used to compute the fisher exact test between the differen country pairs:

```{r}
all_fisher_results <- compute_fisher_results(
  data        = genomes_country,
  country_var = "Country",
  outcome_var = "endometriosis_associated",
  pairs       = pairs
)

all_fisher_results
```

```{r}
#| echo: false
#| eval: false
readr::write_tsv(all_fisher_results, "../results/fisher_results.tsv")
```

The fisher exact tests showed a significant difference in the proportion of endometriosis associated between USA and China with a p-value on 0.02. Usa is having a 2.7 fold higher odds of endometriosis associated genomes.

No significant differences were observed between USA and Denmark (p = 0.155) or between Denmark and China (p = 1). These non-significant results are likely due to the very small number of Danish and Chinese genomes, resulting in wide confidence intervals and low statistical power.
